src/common/transformations/src/transformations/common_optimizations/shared_ops_optimization.cpp

```diff
@@ -0,0 +1,142 @@
+//
+
+#include <openvino/core/validation_util.hpp>
+#include <openvino/op/concat.hpp>
+#include <openvino/op/gather_elements.hpp>
+#include <openvino/op/reshape.hpp>
+#include <openvino/op/slice.hpp>
+#include <openvino/op/tile.hpp>
+#include <openvino/op/util/sub_graph_base.hpp>
+#include <transformations/common_optimizations/shared_ops_optimization.hpp>
+
+#include "itt.hpp"
+
+using namespace std;
+using namespace ov;
+using namespace ov::op;
+
+bool shared_node_optimization(const shared_ptr<Model>& model,
+                              const unordered_map<Node::type_info_t, bool (*)(const Node*, const Node*)>& rules) {
+    bool rewritten = false;
+
+    for (const auto& op : model->get_ordered_ops()) {
+        // Recursively apply transformation for sub-graph based operations
+        if (auto multi_subgraph_op = dynamic_pointer_cast<op::util::MultiSubGraphOp>(op)) {
+            for (size_t i = 0; i < multi_subgraph_op->get_internal_subgraphs_size(); i++) {
+                if (auto sub_graph = multi_subgraph_op->get_function(i))
+                    rewritten |= shared_node_optimization(sub_graph, rules);
+            }
+        }
+        for (auto& output : op->outputs()) {
+            const auto& target_inputs = output.get_target_inputs();
+            if (target_inputs.size() <= 1)
+                continue;  // nothing to optimize
+            unordered_map<Node::type_info_t, vector<Node*>> type_to_node;
+            for (const auto& input : target_inputs) {
+                auto node = input.get_node();
+                if (node && rules.count(node->get_type_info()))
+                    type_to_node[node->get_type_info()].push_back(node);
+            }
+            for (auto& item : type_to_node) {
+                const auto& shared_nodes = item.second;
+                if (shared_nodes.size() < 2)
+                    continue;
+                const auto& ops_type = item.first;
+                const auto& are_equal = rules.at(ops_type);
+
+                std::vector<bool> visited_nodes(shared_nodes.size(), false);
+                for (size_t i = 0; i < visited_nodes.size(); ++i) {
+                    if (visited_nodes[i])
+                        continue;
+                    const auto& root_op = shared_nodes[i];
+                    visited_nodes[i] = true;
+                    for (size_t j = i + 1; j < visited_nodes.size(); ++j) {
+                        if (visited_nodes[j])
+                            continue;
+                        const auto& child_op = shared_nodes[j];
+                        if (are_equal(root_op, child_op)) {
+                            rewritten |= replace_output_update_name(child_op->output(0), root_op->output(0));
+                            visited_nodes[j] = true;
+                        }
+                    }
+                }
+            }
+        }
+    }
+    return rewritten;
+}
+
+bool inputs_from_same_source_or_equal_constants(const Node* lhs, const Node* rhs) {
```

Comment:
do we have a constant comparator function somewhere in common place? op::utils?
Should we add ==operator to Constant class?

Reply:
It's not there yet. I made it common, but it's in on-hold PR #17694:
[src/common/transformations/include/transformations/utils/utils.hpp](https://github.com/openvinotoolkit/openvino/pull/17694/files#diff-3919034fc619fa9085663dfb6fbf07c63946f27a4d916d3a4fb49356d3e9b9c3R226)
[src/common/transformations/src/transformations/utils/utils.cpp](https://github.com/openvinotoolkit/openvino/pull/17694/files#diff-e6fd768fedc8b143cfb95dfe0eefd599ed0cc5e3455767955d64707819bc7ef4R345-R357)

Reply:
Since this isn't available yet -- I resolve the comment